cmd/tiller-proxy/proxy/proxy.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// LogReleaseStatus prints the status of the given release if exists
+func (p *Proxy) LogReleaseStatus(relName string) {
+	status, err := p.helmClient.ReleaseStatus(relName)
+	if err == nil {
+		if status.Info != nil && status.Info.Status != nil {
+			log.Printf("Release status: %s", status.Info.Status.Code)
+		}
+	} else {
+		log.Printf("Unable to fetch release status for %s: %v", relName, err)
+	}
+}
+
+// ListReleases list releases in a specific namespace if given
+func (p *Proxy) ListReleases(namespace string) ([]AppOverview, error) {
+	list, err := p.helmClient.ListReleases()
+	if err != nil {
+		return []AppOverview{}, fmt.Errorf("Unable to list helm releases: %v", err)
+	}
+	appList := []AppOverview{}
+	for _, r := range list.Releases {
+		if namespace == "" || namespace == r.Namespace {
+			appList = append(appList, AppOverview{r.Name, r.Chart.Metadata.Version, r.Namespace})
+		}
+	}
+	return appList, nil
+}
+
+func lock(name string) {
+	if appMutex[name] == nil {
+		appMutex[name] = &sync.Mutex{}
+	}
+	appMutex[name].Lock()
+}
+
+func unlock(name string) {
+	appMutex[name].Unlock()
+}
+
+// CreateRelease creates a tiller release
+func (p *Proxy) CreateRelease(namespace string, rawRelease []byte) (*release.Release, error) {
+	hrelease := &helmRelease{}
+	err := json.Unmarshal(rawRelease, hrelease)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse request body: %v", err)
+	}
+	lock(hrelease.ReleaseName)
+	defer unlock(hrelease.ReleaseName)
+	ch, err := p.getChart(hrelease)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to retrieve chart info: %v", err)
+	}
+	if err != nil {
+		return nil, err
+	}
+	rel, err := p.deploy(hrelease.ReleaseName, namespace, hrelease.Values, ch)
+	if err != nil {
+		return nil, fmt.Errorf("Unable create release: %v", err)
+	}
+	return rel, nil
+}
+
+// UpdateRelease upgrades a tiller release
+func (p *Proxy) UpdateRelease(name, namespace string, rawRelease []byte) (*release.Release, error) {
+	lock(name)
+	defer unlock(name)
+	var rel *release.Release
+	// Check if the release already exists
+	_, err := p.get(name, namespace)
+	if err != nil && isNotFound(err) {
+		return nil, fmt.Errorf("Release %s not found in the namespace %s. Unable to update it", name, namespace)
+	} else if err != nil {
+		return nil, err
+	} else {
+		hrelease := &helmRelease{}
+		err := json.Unmarshal(rawRelease, hrelease)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to parse request body: %v", err)
+		}
+		ch, err := p.getChart(hrelease)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to retrieve chart info: %v", err)
+		}
+		if err != nil {
+			return nil, err
+		}
+		rel, err = p.update(hrelease.ReleaseName, namespace, hrelease.Values, ch)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to update release: %v", err)
+		}
+	}
+	return rel, nil
+}
+
+// GetRelease returns the info of a release
+func (p *Proxy) GetRelease(name, namespace string) (*release.Release, error) {
+	lock(name)
```

Comment:
since this is read-only, do we need the lock? I think we only care about e.g. deleting a release whilst it's still being created?

Reply:
it's not critical but it's something nice to have. We could be returning the wrong version or a missing release.